### ISSUE TITLE
MouseTracker pass drag/pinch events through overlay to canvas?

### ISSUE THREAD

**Issue by wesleyfok** [openseadragon/openseadragon#1652]
We're using overlays to display content on top of an OpenSeadragon image viewer, and have set up a MouseTracker instance to handle click/tap events. All this works fine; the problem is that any OTHER events that start on the overlay, like a drag or a pinch move, get absorbed by the overlay and never make it to OpenSeadragon proper. This means that if a user is using a touchscreen device and starts a drag or a pinch on top of an overlay, the image doesn't move or zoom as expected. The only touch points that register properly are ones that start outside the overlay area (so ex. if you pinch to zoom and one of your touch points is outside the overlay, OSD translates this into a drag event and pans the image instead).

Ideally, I want the overlay to act as if it's not there at all if a user tries to pinch/drag on top of an overlay. But I can't find any way to pass the event through the overlay to OSD; instead the overlay stops propagation of other events completely.

**Comment by wesleyfok** [comment]
Quick update: I tried the hackish workaround documented in https://github.com/openseadragon/openseadragon/issues/1586#issuecomment-459732136 and it actually does work, but only partially. Drag events work fine, and pinch events work fine as well if both touch points are in the same overlay. If the touchpoints are in different overlays or if one is on the viewer surface and the other is on the overlay, the event gets interpreted as a drag instead.

**Comment by iangilman** [comment]
@msalsbery Any thoughts?

**Comment by iangilman** [comment]
@wesleyfok can you share a repo, on CodePen or something?

**Comment by wesleyfok** [comment]
@iangilman The JSFiddle in the ticket I referenced above has a good reproduction of the original issue: http://jsfiddle.net/EQz34/70/

The red and yellow overlays have MouseTrackers attached; the blue one does not. Attempting to drag or pinch the canvas with one or both fingers on either the red or yellow overlays (or both) fails. The blue one does not block normal touch events and thus dragging or pinching with one or both fingers on the blue overlay works fine.

I don't have a CodePen for the hack suggested in the previous ticket, as it requires modification of the base OpenSeadragon library, but I can try to put one together.

**Comment by iangilman** [comment]
Thank you!

Well, I think the ultimate fix here is to revamp the mouse tracker so it doesn't swallow mouse events. That would allow more flexible event handling. 

Short of that, one approach would be to lose the MouseTrackers and just rely on the canvas events. That would involve doing your own hit detection. Here's an example: https://codepen.io/iangilman/pen/byoqXE

**Comment by msalsbery** [comment]
Working on this now!

**Comment by wesleyfok** [comment]
@msalsbery Oooh, thanks! We'll try to incorporate the new version into our site when we get a chance.

**Comment by msalsbery** [comment]
@wesleyfok  I wouldn't release it to the world yet, as we're still testing and it hasn't been merged in. ;-)

But if you get a little time to try it on your site(s) I would really appreciate any feedback!

[Preview builds and demos here](https://github.com/openseadragon/openseadragon/pull/1872#issuecomment-673608460)

**Comment by msalsbery** [comment]
@wesleyfok  Also, now that I look at this issue, there may be code changes you'll need to work with propagation of events between overlays and viewers. I can definitely work with you on ways to accomplish that. Hopefully this patch makes that stuff easier, and the necessary changes won't be extreme and won't be much different than what needs to be done with event propagation on any site.

**Comment by msalsbery** [comment]
With v2.5.0, you can add something like this
```
                preProcessEventHandler: (event) => {
                		event.preventGesture = true;
                }
```
to your overlay MouseTrackers. 

The preProcessEventHandler callback is called for several events before the MouseTracker processes them. The `eventType` property in the passed event object can be used to determine which event is being processed: `"gotpointercapture", "lostpointercapture", "pointerenter", "pointerleave", "pointerover", "pointerout", "pointerdown", "pointerup", "pointermove", "pointercancel", "wheel"`

The `preventGesture` flag is unique to OpenSeadragon. 

There's a few other flags that can be used for finer-grained control of event propagation and default event handling, the most useful are `stopPropagation` and `preventDefault`

**Comment by wesleyfok** [comment]
I think to make this work on our site back when I posted this issue, we ended up using a workaround from earlier in the thread, so we'd probably want to back that out before we looked at integrating v2.5.0 anyways. It'll probably take a bit of time on our end but just wanted to pass on my appreciation in the meantime!

**Comment by msalsbery** [comment]
@wesleyfok

Fixed for v2.5.0 (#1872)

[Here's a CodePen](https://codepen.io/msalsbery/pen/dyOeXqO) that shows what can be done...

Feel free to reopen if issue persists!
